File: frontend/src/types/subject.ts
export interface Topic {
  id: string
  subject_id: string
  title: string
  description: string | null
  is_completed: boolean
  order_index: number
  created_at: string
  updated_at: string
}

export type SubjectPriority = 'low' | 'medium' | 'high'

export interface Subject {
  id: string
  name: string
  description: string | null
  color: string | null
  priority: SubjectPriority
  topic_count: number
  completed_topic_count: number
  progress_percentage: number
  created_at: string
  updated_at: string
}

export interface SubjectDetail extends Subject {
  topics: Topic[]
}

export interface SubjectCreateRequest {
  name: string
  description?: string | null
  color?: string | null
  priority?: SubjectPriority
}

export interface SubjectUpdateRequest {
  name?: string
  description?: string | null
  color?: string | null
  priority?: SubjectPriority
}

export interface TopicCreateRequest {
  title: string
  description?: string | null
  order_index?: number
}

export interface TopicUpdateRequest {
  title?: string
  description?: string | null
  is_completed?: boolean
  order_index?: number
}

File: backend/app/models/subject.py
import uuid
from datetime import datetime, timezone
from typing import TYPE_CHECKING

from sqlalchemy import DateTime, ForeignKey, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base

if TYPE_CHECKING:
    from app.models.topic import Topic
    from app.models.user import User


class Subject(Base):
    __tablename__ = "subjects"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    name: Mapped[str] = mapped_column(String(150), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    color: Mapped[str | None] = mapped_column(String(20), nullable=True)
    priority: Mapped[str] = mapped_column(String(10), nullable=False, default="medium")
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    owner: Mapped["User"] = relationship(back_populates="subjects")
    topics: Mapped[list["Topic"]] = relationship(
        back_populates="subject", cascade="all, delete-orphan", order_by="Topic.order_index"
    )

    @property
    def topic_count(self) -> int:
        return len(self.topics)

    @property
    def completed_topic_count(self) -> int:
        return sum(1 for topic in self.topics if topic.is_completed)

    @property
    def progress_percentage(self) -> int:
        if not self.topics:
            return 0
        return round((self.completed_topic_count / self.topic_count) * 100)